### PR TITLE
Use ip_unprivileged_port_start=53 for coredns

### PR DESCRIPTION
**Pull request by pacoxu**
This is a safe sysctl since 1.22 kubelet. I have tested it in the 1.22 cluster.

/kind cleanup

xref #56374 and #102612

```release-note
kubeadm: use safe `net.ipv4.ip_unprivileged_port_start=53` instead of NET_BIND_SERVICE capability.
```

Since kubeadm should support -2 version kubelet, we may use `ip_unprivileged_port_start` **later in 1.24 or later release**.
I am not quite sure if we should move to `ip_unprivileged_port_start`?

This is supported both in containerd(https://github.com/containerd/containerd/pull/6170 since 1.6) and cri-o (https://github.com/cri-o/cri-o/issues/6706 always)

> 
> -  It has been a safe sysctl since in Kubernetes v1.22. https://github.com/kubernetes/kubernetes/issues/103298
> -  containerd supports the option in CRI since containerd v1.6: https://github.com/containerd/containerd/pull/6170
> - - and containerd plans to make it default in v2.0 https://github.com/containerd/containerd/issues/6924
> 
> https://github.com/containerd/containerd/blob/main/RELEASES.md#kubernetes-support
> 
> 
> 
> Kubernetes Version | containerd Version | CRI Version
> -- | -- | --
> 1.20 | 1.5.0+ | v1alpha2
> 1.21 | 1.5.0+ | v1alpha2
> 1.22 | 1.5.5+ | v1alpha2
> 1.23 | 1.6.0+, 1,5.8+ | v1, v1alpha2
> 1.24 | 1.6.4+, 1.5.11+ | v1, v1alpha2
> 1.25 | 1.6.4+, 1.5.11+ | v1, v1alpha2 **
> 1.26 | 1.6.tbd, 1.7.tbd | v1
> 
> 
> Since v1.26, containerd should use containerd 1.6+. I think we can merge this in favor of all above.

### CODE REVIEW

**Comment by neolit123** [comment]
not so sure about this change, so i'd request review from people who would better understand the impact.

> Since kubeadm should support -2 version kubelet

kubeadm supports kubelet N and N-1, but not N-2.

> we may use ip_unprivileged_port_start later in 1.24 or later release.
I am not quite sure if we should move to ip_unprivileged_port_start?

this is more of a linux net stack question, but why is `ip_unprivileged_port_start` confusingly under `net.ipv4`, what about `ipv6`?

> ip_unprivileged_port_start=53

is `53` always expected?

(from the linked issue):
> securityContext.sysctls instead of container securityContext.capabilities(NET_BIND_SERVICE).

we are already dropping `all` while only granting it `NET_BIND_SERVICE`.
why is the sysctl preferred here?

File: cmd/kubeadm/app/phases/addons/dns/manifests.go
Comment (vinayakankugoyal) [inline cmd/kubeadm/app/phases/addons/dns/manifests.go:82]: I like the change but the whole point of this is that the containers can now be unprivileged? So can we set change the UID and GID we run this as, using runAsUser and runAsGroup?

File: cmd/kubeadm/app/phases/addons/dns/manifests.go
Comment (vinayakankugoyal) [inline cmd/kubeadm/app/phases/addons/dns/manifests.go:131]: This container now does not need any capabilities they can all be dropped:
```
securityContext:
  capabilities:
    drop:
    - ALL
```

**Review by vinayakankugoyal — CHANGES_REQUESTED** [review]

File: cmd/kubeadm/app/phases/addons/dns/manifests.go
Comment (neolit123) [reply cmd/kubeadm/app/phases/addons/dns/manifests.go:82]: /cc @chrisohaver @rajansandeep

do you think we can switch the kubeadm coredns pods to non-0 GID/UID?

Comment (vinayakankugoyal) [reply cmd/kubeadm/app/phases/addons/dns/manifests.go:82]: Yes, we can.

Comment (neolit123) [reply cmd/kubeadm/app/phases/addons/dns/manifests.go:82]: yes, i have assumed so.
pinged the coredns maintainers to see if they are confident about the `net.ipv4.ip_unprivileged_port_start` change and this PR in general.

**Comment by neolit123** [comment]
@vinayakankugoyal note this edit from my comment above:
> EDIT: i guess supporting ambient caps would allows us to continue using it instead of the sysctl solution?

instead of net.ipv4.ip_unprivileged_port_start, shouldn't we wait for ambient caps to be enabled for Pods?
https://github.com/kubernetes/enhancements/pull/2757

File: cmd/kubeadm/app/phases/addons/dns/manifests.go
Comment (chrisohaver) [reply cmd/kubeadm/app/phases/addons/dns/manifests.go:82]: I ... need to research this to understand it better before I can weigh in.

Comment (chrisohaver) [reply cmd/kubeadm/app/phases/addons/dns/manifests.go:82]: I don't foresee any issues with this.

Comment (pacoxu) [reply cmd/kubeadm/app/phases/addons/dns/manifests.go:82]: @vinayakankugoyal maybe we can use `sysctl` and non-root user for coredns if RootlessControlPlane is enabled.

File: cmd/kubeadm/app/phases/addons/dns/manifests.go
Comment (pacoxu) [reply cmd/kubeadm/app/phases/addons/dns/manifests.go:131]: Updated.

**Comment by pacoxu** [comment]
> instead of net.ipv4.ip_unprivileged_port_start, shouldn't we wait for ambient caps to be enabled for Pods? [kubernetes/enhancements#2757](https://github.com/kubernetes/enhancements/pull/2757)

```
      securityContext:
              runAsUser: 1000
              runAsGroup: 1000
              capabilities:
                      drop: ["All"]
                      add: ["NET_BIND_SERVICE"]
      image: simpleserver:filecaps
      command: ["simpleserver", "--port", "80"]
```

For rootless, this may be more general to understand.

File: cmd/kubeadm/app/phases/addons/dns/manifests.go
Comment (neolit123) [inline cmd/kubeadm/app/phases/addons/dns/manifests.go:140]: I don't think we should be running this with an arbitrary user / group until we reserve them the same way we did for the control plane - RootlessControlPlane.

There is also a binding problem here. If coredns in kubeadm moves to an operator the user / group reservation must be propageted to the coredns container that the operator manages.

Comment (pacoxu) [reply cmd/kubeadm/app/phases/addons/dns/manifests.go:140]: Just an idea. We may do some alpha support for this in 1.24.

RootlessControlPlane is still alpha now. And we may add this and reserve the user/group for coredns like control-plane and using this safe sysctl if RootlessControlPlane is enabled in 1.24.

In the future release when we want to graduate RootlessControlPlane to beta, we can get more feedback during that time.

Comment (neolit123) [reply cmd/kubeadm/app/phases/addons/dns/manifests.go:140]: There is a kep in k/enhancements for Linux "user namespaces" which from my understanding renders our kubeadm manual user/group assignment redundant.

For now the plan was to graduate the rootless kubeadm mode to beta in 1.24. But i guess we need to discuss it more..

Comment (vinayakankugoyal) [reply cmd/kubeadm/app/phases/addons/dns/manifests.go:140]: AFAIK the user namespaces kep alpha only applies to containers that don't mount hostVolumes.

**Comment by dims** [comment]
this PR has been on hold for about 2 weeks. Can you please see how to unblock this by reviewing comments etc? or please close this out (Trying to declutter our open list of PRs)

**Comment by dims** [comment]
There is a cncf-cla:no label on this PR. Can you please see the bot comments above to fix the CLA? This PR cannot progress unless the bot is happy with the CLA situation

File: cmd/kubeadm/app/phases/addons/dns/manifests.go
Comment (pacoxu) [reply cmd/kubeadm/app/phases/addons/dns/manifests.go:140]: [User namespace](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/127-user-namespaces#phase-2-pods-with-volumes
) phase 1 is ready in v1.25 for pod.
- Phase 2: pods with volumes.

Comment (pacoxu) [reply cmd/kubeadm/app/phases/addons/dns/manifests.go]: `RootlessControlPlane` is out of the topic of the initial PR.

**Comment by pacoxu** [comment]
/remove-lifecycle stale
Since the safe sysctl was added almost 2 years, is it better to alter it?

**Comment by pacoxu** [comment]
-  It has been a safe sysctl since in Kubernetes v1.22. https://github.com/kubernetes/kubernetes/issues/103298
-  containerd supports the option in CRI since containerd v1.6: https://github.com/containerd/containerd/pull/6170
- - and containerd plans to make it default in v2.0 https://github.com/containerd/containerd/issues/6924

https://github.com/containerd/containerd/blob/main/RELEASES.md#kubernetes-support



Kubernetes Version | containerd Version | CRI Version
-- | -- | --
1.20 | 1.5.0+ | v1alpha2
1.21 | 1.5.0+ | v1alpha2
1.22 | 1.5.5+ | v1alpha2
1.23 | 1.6.0+, 1,5.8+ | v1, v1alpha2
1.24 | 1.6.4+, 1.5.11+ | v1, v1alpha2
1.25 | 1.6.4+, 1.5.11+ | v1, v1alpha2 **
1.26 | 1.6.tbd, 1.7.tbd | v1


Since v1.26, containerd should use containerd 1.6+. I think we can merge this in favor of all above.

**Comment by neolit123** [comment]
what about other container runtimes?
also, any updates on ambient caps as alternative?
https://github.com/kubernetes/kubernetes/pull/105309#issuecomment-929479297

**Comment by pacoxu** [comment]
> what about other container runtimes?

I only did some tests on containerd. (Only checked containerd. And cri-o may not support it.)


> also, any updates on ambient caps as alternative?
> [#105309 (comment)](https://github.com/kubernetes/kubernetes/pull/105309#issuecomment-929479297)

https://github.com/containerd/containerd/pull/6153 is still not merged on contianerd side.
The only change in K/K is https://github.com/kubernetes/kubernetes/pull/104620 in v1.24 as I know.

**Comment by haircommander** [comment]
CRI-O needs no additional configuration to allow sysctls that come from the Kubelet. if it specifies it from the pod spec, CRI-O will respect it.

**Comment by pacoxu** [comment]
/remove-lifecycle stale

- containerd: kube 1.26 + will support only containerd 1.6+. 
- cri-o: this is supported.

**Comment by neolit123** [comment]
still not sure what to do with this. to my understanding we can add this change only for 1.28 and all CR will support it.

one unanswered Q from above:

> this is more of a linux net stack question, but why is `ip_unprivileged_port_start` confusingly under `net.ipv4`, what about `ipv6`?

**Comment by pacoxu** [comment]
net.ipv4.ip_unprivileged_port_start requires kernel version 4.11 or higher. 
/hold
if I understand correctly, even kube and container runtime supports to configure it, this will not work on lower kernel size.

https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/networking/ip-sysctl.txt?h=v4.11#n832

> Thanks for the great answer! It appears that this functionality first appeared in [Linux 4.11 in April 2017](https://kernelnewbies.org/Linux_4.11), so it wasn't around in 2009 when I first asked this question. I also did a quick test, and it seems to also work for IPV6, even though "ipv4" is in the sysctl name. –

From https://stackoverflow.com/questions/413807/is-there-a-way-for-non-root-processes-to-bind-to-privileged-ports-on-linux#comment90027734_51439516.

**Pull request closed without merging**
